### DISCUSSION TITLE
Can drawings extracted from a PDF be saved as images?

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#1442] [Looking for help] [answered]
Is it possible to extract drawings from a PDF and save them as images? Do drawings in a PDF have xref values?

**Comment by JorjMcKie** [comment, marked as answer]
> Is it possible to extract drawings from a PDF and save them as images?

Yes, indirectly: Use the drawing's rectangle as `clip` parameter and make a pixmap from it: `pix = page.get_pixmap(clip=path["rect"])`.

> Do drawings in a PDF have xref values?

No, they only exist as code snippets inside a page's `/Contents` object.

**Reply by meghanaviyyapu** [reply]
Can you let me know how the order in which  drawings and images present in the PDF can be known when images and drawings are extracted separately?

**Reply by JorjMcKie** [reply]
Please consult the documentation:
* `get_drawings()` dictionaries have a `"seqno"` key which numbers the actions in the `/Contents`.
* Image extractions do not have that, but there is `Page.get_bboxlog()`, a list of rectangles tagged with the type of content placed in each rectangle. So you can match an image's bbox with the index number of its bbox in `get_bboxlog()` to indirectly find out its sequence number.

**Reply by meghanaviyyapu** [reply]
Thank you for providing a detailed explanation. Is there a way to extract only specific drawings from PDF?

**Reply by meghanaviyyapu** [reply]
> Please consult the documentation:
> 
> * `get_drawings()` dictionaries have a `"seqno"` key which numbers the actions in the `/Contents`.
> * Image extractions do not have that, but there is `Page.get_bboxlog()`, a list of rectangles tagged with the type of content placed in each rectangle. So you can match an image's bbox with the index number of its bbox in `get_bboxlog()` to indirectly find out its sequence number.

I have used this method but I have found that although image was present before drawings in the PDF, the sequence number of image was greater than the drawings present in the PDF.

**Reply by JorjMcKie** [reply]
> Is there a way to extract only specific drawings from PDF?

No, use the key-value items of these dictionaries to make your desired selection.

**Comment by JorjMcKie** [comment]
> I have used this method but I have found that although image was present before drawings in the PDF, the sequence number of image was greater than the drawings present in the PDF.

The sequence in which bboxes are encountered in the (list of) `/Contents` objects is what counts. Not the time of the insertion.

**Reply by meghanaviyyapu** [reply]
Okay is there a way to get them on the basis of time of insertion.

**Reply by JorjMcKie** [reply]
no

**Reply by meghanaviyyapu** [reply]
Okay Thanks.
![image](https://user-images.githubusercontent.com/36301306/145386701-2e91d1fe-3b09-4c68-b215-76ad9aecd79e.png)

![image](https://user-images.githubusercontent.com/36301306/145386269-f29c79c1-52f9-4e13-a439-cce864f6c993.png)
Can  you let me know if there is a way to map image xref and Figure xref.

**Reply by JorjMcKie** [reply]
Sorry, I don't know that.

**Reply by meghanaviyyapu** [reply]
Is it possible to extract tables or the image of table from PDF using PyMuPDF?

**Reply by JorjMcKie** [reply]
PyMuPDF does not detect tables.

**Reply by meghanaviyyapu** [reply]
Is there a way to extract regions of interest (widgets and the text surrounding the widgets) from PDF?

**Reply by JorjMcKie** [reply]
Widgets have a `rect` property that you can use.

**Reply by meghanaviyyapu** [reply]
I have used rect property of widgets to draw rectangles around them, but in some cases the rectangle was getting drawn at a different location in the PDF in spite of using the rect value of widget. 

Providing the PDF updated after drawing rectangles around widgets.
[1.pdf](https://github.com/pymupdf/PyMuPDF/files/7708847/1.pdf)

Code used
![image](https://user-images.githubusercontent.com/36301306/145936767-300e77c4-7b27-4966-ab50-ac1e8e62bbec.png)

**Comment by JorjMcKie** [comment]
Everything works just fine.
```python
>>> import fitz
>>> doc=fitz.open("1.pdf")
>>> for page in doc:
            shape=page.new_shape()
            page.clean_contents()
            for w in page.widgets():
                shape.draw_rect(w.rect)
            shape.finish(fill=(1,0,0))
            shape.commit()
>>> doc.ez_save("1-y.pdf")
>>> 
```
Then used some viewer to check the button boxes.
[1-y.pdf](https://github.com/pymupdf/PyMuPDF/files/7710373/1-y.pdf)
Yes, there ar some misplaced rectangles, but they have been there before, only with finer lines.

**Comment by meghanaviyyapu** [comment]
Thanks a lot. This really helps.

**Reply by meghanaviyyapu** [reply]
Can you let me know whether PyMuPDF extracts headers and footers from PDFs?

**Reply by JorjMcKie** [reply]
it does not

**Reply by meghanaviyyapu** [reply]
q 1 0 0 1 548.9082 101.0961 cm
0 0 m
-3.982 -10.174 l
-7.85 0 l
-15.476 0 l
-23.549 -21.966 l
-15.297 -21.966 l
-11.489 -11.954 l
-7.742 -21.966 l
-0.273 -21.966 l
7.837 0 l
h
f
Q
q 1 0 0 1 559.6263 101.0961 cm
0 0 m
-8.115 -21.966 l
-0.409 -21.966 l
1.088 -18.11 l
9.332 -18.11 l
10.839 -21.966 l
18.488 -21.966 l
10.481 0 l
h
5.143 -6.726 m
7.823 -13.352 l
2.542 -13.352 l
h
f
Q
q
507.034 79.096 71.08 22 re
W n
q 1 0 0 1 523.0413 87.1387 cm
0 0 m
0.041 0 2.27 -0.012 2.314 0 c
2.352 0.004 -0.666 -8.009 y
-4.539 -8.051 -5.001 -8.042 v
-5.007 -8.032 l
-9.875 -8.032 -12.765 -5.946 -14.41 -3.36 c
-16.038 -0.783 -16.007 2.215 y
-16.007 5.529 -14.072 9.031 -11.466 11.165 c
-9.092 13.039 -7.312 13.856 -2.998 13.945 c
-2.669 13.957 7.709 13.957 y
4.78 6.282 l
0.756 6.282 l
0.456 6.282 -3.075 6.261 -3.498 6.223 c
-3.787 6.197 -6.336 5.77 -6.783 3.042 c
-7.104 1.092 -5.785 0.111 -4.64 0.085 c
-1.882 0.017 0 0 y
f
Q
EMC 

I want to convert only specific drawings into images. For example, I have identified the above drawing data in the content stream. I want to convert this data to an image. Is it possible to do that using the above data?

**Comment by JorjMcKie** [comment]
You have to convert coordinates given in the contents stream to MuPDF coordinates yourself and then try to match them with coordinates in the `page.get_drawings()` output obviously. For example, `1 0 0 1 548.9082 101.0961 cm` corresponds to a matrix `fitz.Matrix(1, 0, 0, 1, 548.9082, 101.0961)`, which positions to point `fitz.Point(548.9082, 101.0961) * page.transformation_matrix` which will be applied to the commands that follow: commands `0 0 m -3.982 -10.174 l` draw a line (given in PDF coordinates) from `Point(0, 0)` to `Point(-3.982, -10.174)`, etc.
Probably easier to directly look at the drawings output ...

**Reply by meghanaviyyapu** [reply]
Ok Thanks
